File: leetcode/42_trapping_rain_water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    if (!height) return 0;

    let leftMax = [];
    let rightMax = [];
    let totalWater = 0;
    let length = height.length;
    // left max instantiate it as the first height in our array
    leftMax[0] = height[0];
    // right max instantiate it as the last height in our array
    rightMax[length-1] = height[length - 1];


    for (let i = 1; i < length; i++) {
        // the left max at the index is either the max of the height currently or our previous max
        leftMax[i] = Math.max(height[i], leftMax[i - 1]);
    }

    // go from right to left, start at second last index
    for (let j = length - 2; j >= 0; j--) {
        rightMax[j] = Math.max(height[j], rightMax[j + 1]);
    }

    // start at index 1 since at the 0th index the rain cannot be trapped anyways
    for (let i = 1; i < length - 1; i++) {
        // our total water is the minimum of the index of either going from left to right or right to left minus the height of the building
        totalWater += Math.min(leftMax[i], rightMax[i]) - height[i];
    }

    return totalWater;
};

// Time Complexity: O(N)
// Space Complexity: O(N)

File: leetcode/42_trapping_rain_water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        if not height:
            return 0

        total_water = 0
        left = 0
        right = len(height) - 1
        left_max = height[left]
        right_max = height[right]

        while left < right:
            # if our height at the left pointer is less than or equal the height at the right pointer
            if height[left] <= height[right]:
                # water can be trapped
                if height[left] < left_max:
                    total_water += left_max - height[left]
                else:
                    # update our left_max
                    left_max = height[left]
                # move left pointer to the right
                left += 1
            # if our height at the right pointer is less than the height at the left pointer
            else:
                # water can be trapped if the current right pointer is less than the right max
                if height[right] < right_max:
                    # add the difference between the right max and the current height, the SA is one unit
                    total_water += right_max - height[right]
                else:
                    # update our right_max
                    right_max = height[right]
                # move right pointer left
                right -= 1

        return total_water

# Time: O(N + N) => O(N) since we utilize two pointers to iterate through the array of heights
# Space: O(1) since we are not using any extra space